azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java

```diff
@@ -150,54 +164,115 @@ public String toString() {
+            refreshTokenFromResult = (String) FieldUtils.readField(result, "refreshToken", true);
+        } catch (IllegalAccessException e) {
+            throw new AzureToolkitAuthenticationException("Cannot read refreshToken from IAuthenticationResult.");
         }
+
+        return refreshTokenFromResult;
     }
 
-    protected void initializeTenants() {
-        // in azure cli, the tenant ids from 'az account list' should be less/equal than the list tenant api
-        if (this.entity.getTenantIds() == null) {
-            TokenCredential credential = entity.getTenantCredential().createTokenCredential(null);
-            List<String> allTenantIds = listTenantIds(entity.getEnvironment(), credential);
-            this.entity.setTenantIds(allTenantIds);
-        }
+    private Mono<Tuple2<List<String>, AccessToken>> verifyAndGetTenantId(TokenCredential credential) {
+        AzureEnvironment env = ObjectUtils.firstNonNull(this.entity.getEnvironment(), Azure.az(AzureCloud.class).getOrDefault());
+        TokenRequestContext tokenRequestContext = new TokenRequestContext()
+                .addScopes(ScopeUtil.resourceToScopes(env.getManagementEndpoint()));
+        Mono<AccessToken> mono = credential instanceof TenantCredential ?
+                ((TenantCredential) credential).getAccessToken(null, tokenRequestContext) :
+                credential.getToken(tokenRequestContext);
+
+        return mono.flatMap(accessToken ->
+                Mono.zip(AzureResourceManager.authenticate(request -> Mono.just(accessToken), new AzureProfile(env))
+                        .tenants().listAsync().map(Tenant::tenantId).collectList(), Mono.just(accessToken)));
     }
 
-    protected void initializeSubscriptions() {
-        List<Subscription> subscriptions = listSubscriptions(entity.getTenantIds(), entity.getTenantCredential());
-        entity.setTenantIds(subscriptions.stream().map(Subscription::getTenantId).distinct().collect(Collectors.toList()));
-        entity.setSubscriptions(subscriptions);
-        selectSubscriptionInner(subscriptions, this.entity.getSelectedSubscriptionIds());
+    protected Mono<Void> initializeSubscriptions() {
+        return listSubscriptions(entity.getTenantIds(), entity.getTenantCredential()).doOnSuccess(subscriptions -> {
+            entity.setTenantIds(subscriptions.stream().map(Subscription::getTenantId).distinct().collect(Collectors.toList()));
+            entity.setSubscriptions(subscriptions);
+            selectSubscriptionInner(subscriptions, this.entity.getSelectedSubscriptionIds());
 
-        // no subscriptions are selected, selected all
-        if (!subscriptions.isEmpty() && subscriptions.stream().noneMatch(Subscription::isSelected)) {
-            subscriptions.forEach(s -> s.setSelected(true));
-        }
+            // no subscriptions are selected, selected all
+            if (!subscriptions.isEmpty() && subscriptions.stream().noneMatch(Subscription::isSelected)) {
```

Comment:
move logic here to `selectSubscriptionInner`?

Reply:
No need since it is select all, selectSubscriptionInner(subscriptions.keySet()) is complex